ql/termstructures/yield/regulatorytermstructure.hpp

```diff
@@ -0,0 +1,175 @@
+
+/*
+ Copyright (C) 2020 Marcin Rybacki
+
+ This file is part of QuantLib, a free-software/open-source library
+ for financial quantitative analysts and developers - http://quantlib.org/
+
+ QuantLib is free software: you can redistribute it and/or modify it
+ under the terms of the QuantLib license.  You should have received a
+ copy of the license along with this program; if not, please email
+ <quantlib-dev@lists.sf.net>. The license is also available online at
+ <http://quantlib.org/license.shtml>.
+
+ This program is distributed in the hope that it will be useful, but WITHOUT
+ ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ FOR A PARTICULAR PURPOSE.  See the license for more details.
+*/
+
+/*! \file regulatorytermstructure.hpp
+    \brief Ultimate Forward Rate term structure
+*/
+
+#ifndef quantlib_regulatory_term_structure_hpp
+#define quantlib_regulatory_term_structure_hpp
+
+#include <ql/quote.hpp>
+#include <ql/termstructures/yield/zeroyieldstructure.hpp>
+
+namespace QuantLib {
+
+    //! Regulatory term structure
+
+    /*! Dutch regulatory term structure for pension funds with a
+        parametrized extrapolation mechanism designed for
+        discounting long dated liabilities.
+
+        Relevant documentation can be found on the Dutch Central
+        Bank website:
+
+        FTK term structure documentation (Financieel toetsingskader):
+        https://www.toezicht.dnb.nl/binaries/50-212329.pdf
+
+        UFR 2015 term structure documentation:
+        https://www.toezicht.dnb.nl/binaries/50-234028.pdf
+
+        UFR 2019 term structure documentation:
+        https://www.rijksoverheid.nl/documenten/kamerstukken/2019/06/11/advies-commissie-parameters
+
+        This term structure will remain linked to the original
+        structure, i.e., any changes in the latter will be
+        reflected in this structure as well.
+
+        \ingroup yieldtermstructures
+
+        \test
+        - the correctness of the returned zero rates is tested by
+          checking them against reference values obtained
+          from the official source.
+        - extrapolated forward is validated.
+        - rates on the cut-off point are checked against those
+          implied by the base curve.
+        - inspectors are tested against the base curve.
+        - incorrect input for cut-off point should raise an exception.
+        - observability against changes in the underlying term
+          structure and the additional components is checked.
+    */
+
+    class UltimateForwardTermStructure : public ZeroYieldStructure {
+      public:
+        UltimateForwardTermStructure(const Handle<YieldTermStructure>&,
+                                     const Handle<Quote>& lastLiquidForwardRate,
+                                     const Handle<Quote>& ultimateForwardRate,
+                                     Time firstSmoothingPoint,
```

Comment:
Is this specified by the regulator as a period (e.g., "20 years")? If so, mapping it to a time would be best left to the curve because it would depend on its internal day-count convention.  The constructor would then take it as a `Period` instance.

Reply:
Hi Luigi, thank you for your feedback. I agree, having Period instead of Time is a better choice. The regulator specifies this threshold point as a period (at the moment it is 20 years, for instance).  I will change this accordingly.